src/react/src/components/Filters/NestedSelect.jsx

```diff
@@ -0,0 +1,232 @@
+
+    const handleClick = () => {
+        if (isExpanded) {
+            setExpandedGroups(
+                expandedGroups.filter(label => label !== data.label),
+            );
+        } else {
+            setExpandedGroups([...expandedGroups, data.label]);
+        }
+    };
+
+    const handleGroupSelect = () => {
+        const groupOptions = data.options.map(option => option.value);
+        const newSelectedOptions = [
+            ...selectedOptions.filter(opt => !groupOptions.includes(opt.value)),
+            ...data.options,
+        ];
+        setSelectedOptions(newSelectedOptions);
+        setMenuIsOpen(false);
+    };
+
+    return (
+        <components.GroupHeading {...props}>
+            <IconButton
+                onClick={handleClick}
+                className={classes.groupHeadingIconButton}
+            >
+                {isExpanded ? <ExpandMoreIcon /> : <ChevronRightIcon />}
+            </IconButton>
+            <ButtonBase
+                onClick={handleGroupSelect}
+                className={classes.groupHeadingButtonBase}
+            >
+                {data.label}
+            </ButtonBase>
+        </components.GroupHeading>
+    );
+};
+
+const Option = props => {
+    const { data, selectProps } = props;
+    const { expandedGroups } = selectProps;
+    const isVisible = expandedGroups.some(group => data.groupLabel === group);
+
+    return isVisible ? <components.Option {...props} /> : null;
+};
+
+const DropdownIndicator = ({ selectProps: { classes } }) => (
+    <div className={classes.dropdownIndicator}>
+        <ArrowDropDownIcon />
+    </div>
+);
+
+const NestedSelect = ({
+    name,
+    label,
+    optionsData,
+    sectors,
+    updateSector,
+    classes,
+    isSideBarSearch,
+    windowWidth,
+    ...rest
+}) => {
+    const selectFilterStyles = makeSelectFilterStyles(windowWidth);
+    const nestedSelectFilterStyles = makeNestedSelectFilterStyles;
+    const combinedStyles = {
+        ...selectFilterStyles,
+        ...nestedSelectFilterStyles,
+    };
+    const [menuIsOpen, setMenuIsOpen] = useState(false);
+    const [expandedGroups, setExpandedGroups] = useState([]);
+
+    const handleInputChange = inputValue => {
+        if (inputValue.trim() === '') {
+            setExpandedGroups([]);
+            return;
+        }
+
+        const newExpandedGroups = optionsData
+            .filter(group =>
+                group.options.some(option =>
+                    option.label
+                        .toLowerCase()
+                        .includes(inputValue.toLowerCase()),
+                ),
+            )
+            .map(group => group.label);
+
+        setExpandedGroups(newExpandedGroups);
+    };
+
+    const handleSelect = selected => {
+        updateSector(selected);
+    };
+
+    const onMenuOpen = () => {
+        setMenuIsOpen(true);
+    };
+
+    const onMenuClose = () => {
+        setExpandedGroups([]);
+        setMenuIsOpen(false);
+    };
```

Comment:
I believe some parts of this method can be converted into React custom hooks to improve the separation of concerns (presentational and implementation layers) to some extent.
Docs:
1. https://sairys.medium.com/react-separating-responsibilities-using-hooks-b9c90dbb3ab9
2. https://react.dev/learn/reusing-logic-with-custom-hooks

Reply:
Moved the state management logic for the menu and expanded groups into custom hooks.